spanner/api/Spanner.Samples.Tests/GetDatabaseDdlAsyncTest.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Copyright 2021 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Threading.Tasks;
+using Xunit;
+
+[Collection(nameof(SpannerFixture))]
+public class GetDatabaseDdlAsyncTest
+{
+    private readonly SpannerFixture _spannerFixture;
+
+    public GetDatabaseDdlAsyncTest(SpannerFixture spannerFixture)
+    {
+        _spannerFixture = spannerFixture;
+    }
+
+    [Fact]
+    public async Task TestGetDatabaseDdlAsync()
+    {
+        var createSingersTable =
+            @"CREATE TABLE Singers (
+                     SingerId INT64 NOT NULL,
+                     FirstName STRING(1024),
+                     LastName STRING(1024),
+                     ComposerInfo BYTES(MAX),
+                 ) PRIMARY KEY (SingerId)";
+        var createAlbumsTable =
+            @"CREATE TABLE Albums (
+                     SingerId INT64 NOT NULL,
+                     AlbumId INT64 NOT NULL,
+                     AlbumTitle STRING(MAX),
+                 ) PRIMARY KEY (SingerId, AlbumId),
+                 INTERLEAVE IN PARENT Singers ON DELETE CASCADE";
+        await _spannerFixture.RunWithTemporaryDatabaseAsync(_spannerFixture.InstanceId, async databaseId =>
+        {
+            var sample = new GetDatabaseDdlAsyncSample();
+            var statements =
+                await sample.GetDatabaseDdlAsync(_spannerFixture.ProjectId, _spannerFixture.InstanceId, databaseId);
+            Assert.Collection(statements,
+                // Only check the start of the statement, as there is no guarantee on exactly
+                // how Cloud Spanner will format the returned SQL string.
+                statement => Assert.StartsWith("CREATE TABLE Singers", statement),
+                statement => Assert.StartsWith("CREATE TABLE Albums", statement)
```

Comment:
Could we also check if the default leader extra statment is there? Even though it's not explicitly stated in the spec, I feel like that's the intention of this sample. But I see that it might be tricky here given that we create a new database for every sample. Whereas in other languages, because we execute the samples in a certain order, we'd create a database with the extra statement for specifying the leader and then dump the DDL.

So you might just be able to call `CreateDatabaseWithDefaultLeaderAsyncSample()` to create the database instead of `RunWithTemporaryDatabaseAsync()`.

Reply:
Good point. I've added an `ALTER DATABASE` statement to the temporary database definition and added asserts for it in the test.

Reply:
We could call `CreateDatabaseWithDefaultLeaderAsyncSample()`, but I would prefer not to for the following two reasons:
1. That sample is specifically for creating a database with a default leader. The fact that it also creates two tables is not really an integral part of the sample (there's nothing against that sample just creating an empty database). If anyone would change that sample in the future to create a different data model, or no tables at all, this test would start failing. I don't think that kind of cross-dependence between unrelated tests and samples is a good thing.
2. See also [my comment here](https://github.com/GoogleCloudPlatform/dotnet-docs-samples/pull/1366#discussion_r680315614). By creating (and dropping) a single database, the test case is kept independent of all other tests and samples, and could easily also be run without all the initialization in `SpannerFixture`. Running only this test (including the creating and dropping of the database) takes approximately 5-7 seconds. The initialization of `SpannerFixture` takes 4 minutes when it has already been executed at least once, and most of the resources have already been created. That makes the development of this kind of simple samples a lot more of a hassle than it needs to be. It also sometimes makes it very difficult to determine exactly in what the state of a database and/or table is at the moment that a test is being executed. See for example [this test for one of the tagging samples](https://github.com/GoogleCloudPlatform/dotnet-docs-samples/blob/20102071d977d57b3f1ce00da77a54d97df2f6b3/spanner/api/Spanner.Samples.Tests/TransactionTagAsyncTest.cs#L34). I have absolutely no idea what the contents is of the `Venues` table at the moment that this test is being executed, as it might have been modified by any of the preceding sample tests.

Reply:
OK on point 1. (I've never been a fan of using different samples for testing another one but we started like that so, .... and I do get that it avoid some code duplication.)

On point 2, see [my comment](https://github.com/GoogleCloudPlatform/dotnet-docs-samples/pull/1366#discussion_r683605984) earlier.